### PR TITLE
feat(ckbtc): Use the new KYT canister in ckbtc deposit flow 

### PR DESCRIPTION
**Pull request by ninegua**
Switch to the new KYT canister in the ckBTC deposit flow. 

Also enable real https outcalls in system tests with a SSL reverse proxy + bitcoind regtest network.

### CODE REVIEW

File: rs/tests/ckbtc/BUILD.bazel [hunk 1]
```diff
@@ -56,12 +55,33 @@ rust_library(
...
+oci_tar(
+    name = "minica.tar",
+    image = "@minica",
+    repo_tags = ["minica:image"],
+)
+
+uvm_config_image(
+    name = "ckbtc_uvm_config_image",
+    srcs = [
+        ":minica.tar",
```
Comment (basvandijk) [inline rs/tests/ckbtc/BUILD.bazel:67]: What about moving the equivalent [`//rs/tests/networking/canister_http:minica.tar`](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/tests/networking/canister_http/BUILD.bazel?L28-32) to [`/rs/tests/BUILD.bazel`](https://sourcegraph.com/github.com/dfinity/ic/-/blob/rs/tests/BUILD.bazel) and removing this one?

File: rs/tests/ckbtc/src/lib.rs [hunk 2]
```diff
@@ -141,11 +106,62 @@ docker run  --name=bitcoind-node -d \
...
+docker load -i /config/minica.tar
+docker run -v "$(pwd)":/output minica:image -ip-addresses="{btc_node_ipv6}"
+sudo mv "{btc_node_ipv6}"/cert.pem localhost.crt
+sudo mv "{btc_node_ipv6}"/key.pem localhost.key
+sudo chmod 644 localhost*
+
+# Run nginx auto proxy
+docker run -d --name=proxy -e ENABLE_IPV6=true -e DEFAULT_HOST=localhost -p 80:80 -p {HTTPS_PORT}:443 \
+           -v /tmp/certs:/etc/nginx/certs -v /var/run/docker.sock:/tmp/docker.sock:ro \
+           nginxproxy/nginx-proxy
```
Comment (basvandijk) [inline rs/tests/ckbtc/src/lib.rs:131]: Note that since this runs in a fresh Universal VM it will have to download `nginxproxy/nginx-proxy` for every run. This will very soon hit the docker hub rate limit and cause these tests to start failing.

Could you move this image into the UVM config image like you do for minica above?

File: rs/tests/ckbtc/src/lib.rs [hunk 3]
```diff
@@ -141,11 +106,62 @@ docker run  --name=bitcoind-node -d \
...
+    regtest=1
+    debug=1
+    whitelist=::/0
+    fallbackfee=0.0002
+    rpcauth=$BITCOIND_RPC_AUTH
+END
+docker run  --name=bitcoind-node -d \
+  -e VIRTUAL_HOST=localhost -e VIRTUAL_PORT={BITCOIND_RPC_PORT} -v /tmp:/bitcoin/.bitcoin \
+  -p {BITCOIN_CLI_PORT}:{BITCOIN_CLI_PORT} -p {BITCOIND_RPC_PORT}:{BITCOIND_RPC_PORT} \
+  ghcr.io/dfinity/bitcoind@sha256:17c7dd21690f3be34630db7389d2f0bff14649e27a964afef03806a6d631e0f1
```
Comment (basvandijk) [inline rs/tests/ckbtc/src/lib.rs:147]: Ideally this image is also moved in the UVM config image so that it doesn't have to be downloaded each time.

File: rs/bitcoin/ckbtc/minter/src/updates/update_balance.rs [hunk 4]
```diff
@@ -283,60 +283,45 @@ pub async fn update_balance(
...
-                Ok((
-                    response.external_id,
-                    UtxoCheckStatus::Clean,
-                    response.provider,
-                ))
-            }
+        CheckTransactionResponse::Failed(addresses) => {
+            log!(
+                P0,
+                "Discovered a tainted UTXO {} (due to input addresses {})",
```
Comment (gregorydemay) [inline rs/bitcoin/ckbtc/minter/src/updates/update_balance.rs:309]: nit: should we also log the `UpdateBalanceArgs` (principal and subaccount) that triggered this? I have the impression this would be useful for debuggability .

File: rs/bitcoin/ckbtc/minter/src/updates/update_balance.rs [hunk 5]
```diff
@@ -283,60 +283,45 @@ pub async fn update_balance(
...
-        Err(KytError::TemporarilyUnavailable(reason)) => {
+        CheckTransactionResponse::Passed => Ok(UtxoCheckStatus::Clean),
+        CheckTransactionResponse::Unknown(status) => {
             log!(
                 P1,
-                "The KYT provider is temporarily unavailable: {}",
-                reason
+                "The KYT provider is temporarily unavailable: {:?}",
+                status
             );
```
Comment (gregorydemay) [inline rs/bitcoin/ckbtc/minter/src/updates/update_balance.rs:321]: If I recall correctly in that case it could be that the transaction have a lot of inputs and we need to keep trying with more cycles, no?

File: rs/bitcoin/ckbtc/minter/src/state.rs [hunk 6]
```diff
@@ -463,8 +459,8 @@ impl CkBtcMinterState {
             min_confirmations,
             mode,
             new_kyt_principal,
-            kyt_principal,
             kyt_fee,
+            ..
```
Comment (gregorydemay) [inline rs/bitcoin/ckbtc/minter/src/state.rs:463]: nit: this I think I would not add. I think it's a nice feature that if something changes in `UpgradeArgs`, this place will not compile and one needs to actively need to think of the changes that are required in that method.

File: rs/bitcoin/ckbtc/minter/src/state.rs [hunk 7]
```diff
@@ -434,8 +431,8 @@ impl CkBtcMinterState {
             min_confirmations,
             mode,
             kyt_fee,
-            kyt_principal,
             new_kyt_principal,
+            ..
```
Comment (gregorydemay) [inline rs/bitcoin/ckbtc/minter/src/state.rs:435]: nit: same comment as for `upgrade` below.

File: rs/bitcoin/ckbtc/minter/src/state.rs [hunk 8]
```diff
@@ -382,7 +379,7 @@ pub struct CkBtcMinterState {
     pub owed_kyt_amount: BTreeMap<Principal, u64>,
 
     /// A cache of UTXO KYT check statuses.
-    pub checked_utxos: BTreeMap<Utxo, (String, UtxoCheckStatus, Principal)>,
+    pub checked_utxos: BTreeMap<Utxo, (Option<String>, UtxoCheckStatus, Option<Principal>)>,
```
Comment (gregorydemay) [inline rs/bitcoin/ckbtc/minter/src/state.rs:382]: nit: The type of the values is a bit hard to understand, since the semantic is not apparent from the type (e.g. `String` is actually meant to represent a UUID to identify a provider). Maybe it would be better to have this as a proper `struct`? If yes, also ok obviously to do this in a separate PR.

**Review by gregorydemay — COMMENTED** [review]
Thanks a lot @ninegua for this PR! I have a couple of minor comments but nothing critical.

File: rs/bitcoin/ckbtc/minter/src/updates/update_balance.rs (hunk 5 above)
Comment (ninegua) [reply rs/bitcoin/ckbtc/minter/src/updates/update_balance.rs]: Yes, you are right. Good catch! I added retry handling. Please have a look again. Thanks!

File: rs/bitcoin/ckbtc/minter/tests/tests.rs [hunk 9]
```diff
@@ -327,13 +296,8 @@ fn test_upgrade_restricted() {
 
     // upgrade
     let upgrade_args = UpgradeArgs {
-        retrieve_btc_min_amount: Some(2000),
-        min_confirmations: None,
-        max_time_in_queue_nanos: Some(100),
         mode: Some(Mode::RestrictedTo(vec![authorized_principal])),
-        kyt_fee: None,
-        new_kyt_principal: Some(CanisterId::from(0)),
```
[the remark this answers is not in the document]
Comment (ninegua) [reply rs/bitcoin/ckbtc/minter/tests/tests.rs:335]: kyt_principal is effectively being removed, and no longer available in the internal state. The next PR after this will rename `new_kyt_principal` to `kyt_principal`.

File: rs/tests/ckbtc/src/lib.rs [hunk 10]
```diff
@@ -390,7 +405,7 @@ pub async fn install_minter(
         min_confirmations: Some(BTC_MIN_CONFIRMATIONS as u32),
```
[the remark this answers is not in the document]
Comment (ninegua) [reply rs/tests/ckbtc/src/lib.rs:405]: I think it is a real setup that will generate its own threshold ECDSA key. So the `TEST_KEY_LOCAL` can be any string.

File: rs/bitcoin/ckbtc/minter/src/updates/update_balance.rs [hunk 11]
```diff
@@ -297,37 +307,58 @@ async fn kyt_check_utxo(utxo: &Utxo) -> Result<UtxoCheckStatus, UpdateBalanceErr
...
+                    error_code: ErrorCode::KytError as u64,
+                    error_message: format!("KYT error: {:?}", error),
+                });
+            }
         }
     }
+    Err(UpdateBalanceError::GenericError {
+        error_code: ErrorCode::KytError as u64,
+        error_message: format!("The KYT canister required too many calls to check_transaction"),
+    })
```
Comment (gregorydemay) [inline rs/bitcoin/ckbtc/minter/src/updates/update_balance.rs]: Thanks for the fix! 2 follow-up questions:

1. In that case a user would need to call `update_balance` again to retry, correct?
2. Would it be possible to have a pocket IC test for this scenario (minter needs to try at least twice to validate a transaction)?

File: rs/tests/ckbtc/src/lib.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (gregorydemay) [reply rs/tests/ckbtc/src/lib.rs:405]: > So the TEST_KEY_LOCAL can be any string.

No, there are only 3 values (see https://internetcomputer.org/docs/current/references/samples/rust/threshold-ecdsa/#update-source-code-with-the-right-key-id)
> There are three options:
> - dfx_test_key: a default key ID that is used in deploying to a local version of IC (via IC SDK).
> - test_key_1: a master test key ID that is used in mainnet.
> - key_1: a master production key ID that is used in mainnet.


Only `test_key_1` and `key_1` corresponds to a production setting (where an iDKG protocol would be run).

File: rs/bitcoin/ckbtc/minter/src/updates/update_balance.rs [hunk 12]
```diff
@@ -283,60 +283,45 @@ pub async fn update_balance(
...
+        s.new_kyt_principal
+            .expect("BUG: upgrade procedure must ensure that the new KYT principal is set")
             .get()
             .into()
     });
 
-    if let Some((uuid, status, api_key_owner)) = read_state(|s| s.checked_utxos.get(utxo).cloned())
-    {
-        return Ok((uuid, status, api_key_owner));
+    if let Some((_, status, _)) = read_state(|s| s.checked_utxos.get(utxo).cloned()) {
```
[the remark this answers is not in the document]
Comment (gregorydemay) [reply rs/bitcoin/ckbtc/minter/src/updates/update_balance.rs]: > Actually it starts to feel like this cache is perhaps no longer needed since the KYT canister also keeps a cache. Can we remove it (in another ticket)?

Yes, I think it makes sense. In particular the KYT canister should decide whether its cache is up-to-date and will automatically loose it upon upgrades.

File: rs/bitcoin/ckbtc/minter/src/state/eventlog.rs [hunk 13]
```diff
@@ -307,19 +320,9 @@ pub fn replay<I: CheckInvariants>(
...
-                                uuid,
-                            )))
-                        }
-                    };
                 state.mark_utxo_checked(
                     utxo,
-                    uuid,
+                    if uuid.is_empty() { None } else { Some(uuid) },
                     UtxoCheckStatus::from_clean_flag(clean),
                     kyt_provider,
```
[the remark this answers is not in the document]
Comment (gregorydemay) [reply rs/bitcoin/ckbtc/minter/src/state/eventlog.rs:327]: Ha I see, thanks for the explanation!

File: rs/bitcoin/ckbtc/minter/src/state/eventlog.rs [hunk 14]
```diff
@@ -332,8 +335,12 @@ pub fn replay<I: CheckInvariants>(
                 amount,
                 ..
             } => {
-                if let Err(Overdraft(overdraft)) = state.distribute_kyt_fee(kyt_provider, amount) {
-                    return Err(ReplayLogError::InconsistentLog(format!("Attempted to distribute {amount} to {kyt_provider}, causing an overdraft of {overdraft}")));
+                if Some(kyt_provider) != previous_kyt_principal {
+                    if let Err(Overdraft(overdraft)) =
+                        state.distribute_kyt_fee(kyt_provider, amount)
```
[the remark this answers is not in the document]
Comment (gregorydemay) [reply rs/bitcoin/ckbtc/minter/src/state/eventlog.rs:340]: > was only required when replaying testnet events that had empty kyt_provider field, in which case the old code was using kyt_principal as a default substitute.

But for testnet we didn't do KYT before, meaning that there shouldn't be any `DistributedKytFee` event for testnet, no? Would it be maybe more readable if we assume that inside this event we are using Mainnet so that we don't need "this hack of remembering `previous_kyt_principal`"?

**Review by gregorydemay — APPROVED** [review]
Thanks @ninegua for the improvements!

File: rs/bitcoin/ckbtc/minter/src/state/eventlog.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (ninegua) [reply rs/bitcoin/ckbtc/minter/src/state/eventlog.rs:340]: Well, the testnet setup still requires a KYT canister that is set to AcceptAll. And kyt_fee is still deducted from deposit or withdrawal transactions. It is just that `kyt_provider` was set to None, and defaults to `kyt_principal`. The previous logic would still produce all events including `CheckedUtxo` and `DistributedKytFee`.

The purpose of the changes here (and above) is to
1. Do not supply a default value to to kyt_provider when it is None.
2. Be compatible when replaying existing mainnet & testnet events.

File: rs/bitcoin/ckbtc/minter/src/updates/update_balance.rs (hunk 4 above)
Comment (ninegua) [reply rs/bitcoin/ckbtc/minter/src/updates/update_balance.rs]: The changes should be in the code. I think github either redirects users to the old version if you follow the linked code above,  or it just won't show related discussion thread if you go to File Changed tag to view the code & comments.

File: rs/tests/ckbtc/src/lib.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (ninegua) [reply rs/tests/ckbtc/src/lib.rs:402]: The external documentation and spec refer to either mainnet or a developer's local dfx environment. But the ECDSA key id was not hard-coded in replica. For systests, the setup configures one or more local subnets and run its own ECDSA key generation, and would use whatever key name it is set to use. For dfx, its own setup configures the id to be called `dfx_test_key`. If you use dfx then you'll have to stick with this id. If you use mainnet, you'll have to use either `test_key_1` or `key_1`.

I just changed this `TEST_KEY_LOCAL` value to a random string, and the tests still passed.

File: rs/bitcoin/ckbtc/minter/src/updates/update_balance.rs (hunk 11 above)
Comment (ninegua) [reply rs/bitcoin/ckbtc/minter/src/updates/update_balance.rs]: > 1. In that case a user would need to call update_balance again to retry, correct?

Either that or the user can contact DFINITY to report this error. It would certainly be an anomaly. 

> 2. Would it be possible to have a pocket IC test for this scenario (minter needs to try at least twice to validate a transaction)?

The KYT canister has unit tests that where the caller would need to call more than once, but not pocket IC based tests. We will need to tweak some setup in order to trigger this scenario, and in particular, a much lower `CHECK_TRANSACTION_CYCLES_REQUIRED` is required. Maybe not in this PR though. I created a Jira ticket to track this https://dfinity.atlassian.net/browse/XC-241

File: rs/tests/ckbtc/src/lib.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (gregorydemay) [reply rs/tests/ckbtc/src/lib.rs:403]: > I just changed this TEST_KEY_LOCAL value to a random string, and the tests still passed.

Ha interesting, thanks for double checking!

File: rs/tests/ckbtc/src/lib.rs (hunk 3 above)
Comment (ninegua) [reply rs/tests/ckbtc/src/lib.rs]: Done.

File: rs/tests/ckbtc/src/lib.rs (hunk 2 above)
Comment (ninegua) [reply rs/tests/ckbtc/src/lib.rs]: Done.

File: rs/tests/ckbtc/src/lib.rs [hunk 15]
```diff
@@ -141,11 +106,62 @@ docker run  --name=bitcoind-node -d \
...
+#      --cert-file /certs/localhost.crt --key-file /certs/localhost.key --port 80
+"#
+        ))
+        .unwrap());
+
     InternetComputer::new()
-        .with_bitcoind_addr(SocketAddr::new(IpAddr::V6(btc_node_ipv6), 18444))
+        .with_bitcoind_addr(SocketAddr::new(IpAddr::V6(btc_node_ipv6), BITCOIN_CLI_PORT))
         .add_subnet(
             Subnet::new(SubnetType::System)
```
[the remark this answers is not in the document]
Comment (ninegua) [reply rs/tests/ckbtc/src/lib.rs:161]: I don't think it would make much difference. Will leave it as is unless there is a strong reason to change.

**Review by basvandijk — APPROVED** [review]
Very nice!

File: rs/tests/ckbtc/src/lib.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (gregorydemay) [reply rs/tests/ckbtc/src/lib.rs:161]: My main reason would be for cycle accounting (ensuring that the values used do work as expected and we are not running in an unexpected out-of-cycles error)
System subnets do not use cycles. Also ok to do this in a separate PR.

File: rs/bitcoin/ckbtc/minter/src/state.rs [hunk 16]
```diff
@@ -259,6 +259,16 @@ impl UtxoCheckStatus {
     }
 }
 
+/// Relevant data for a checked UTXO. The uuid and kyt_provider are kept
+/// backward compatibility reasons. They should be set to `None` since
```
Comment (THLO) [inline rs/bitcoin/ckbtc/minter/src/state.rs:263]:
```suggestion
/// Relevant data for a checked UTXO. The UUID and `kyt_provider` are kept for
/// backward-compatibility reasons. They should be set to `None` since
```

File: rs/bitcoin/ckbtc/minter/src/updates/update_balance.rs [hunk 17]
```diff
@@ -230,9 +230,9 @@ pub async fn update_balance(
             utxo_statuses.push(UtxoStatus::ValueTooSmall(utxo));
             continue;
         }
-        let (uuid, status, kyt_provider) = kyt_check_utxo(caller_account.owner, &utxo).await?;
+        let status = kyt_check_utxo(&utxo).await?;
```
[the remark this answers is not in the document]
Comment (THLO) [reply rs/bitcoin/ckbtc/minter/src/updates/update_balance.rs]: As discussed [here](https://docs.google.com/document/d/18n33mTyvZaaisN8KU4FeliITbrL8zrUCKxv6tVfXPH0/edit?tab=t.0#heading=h.l39l03muiqfo), the cost for making the HTTPS outcalls is passed on to the user in the form of a KYT (or KYBTC) fee.
While the caller must send **40b cycles**, excess cycles are returned, right?
An average cost of `2.5b cycles` per call (i.e., most calls return more than 39b cycles) would imply a cost of **12.5 sats** at a conservative price of 20,000 XDR per bitcoin.

The proposal is to charge **100 sats** as the fee, i.e., 20 times less than what we charge right now. This should easily cover the costs and help to amortize the costs of the free `update_balance` calls.

**Review by THLO — APPROVED** [review]

File: rs/tests/ckbtc/src/lib.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (ninegua) [reply rs/tests/ckbtc/src/lib.rs:161]: I see, makes sense. But it appears non-trivial to setup an app subnet as I briefly tried, something to do with `use_specified_ids_allocation_range` that I don't fully understand yet. Will leave it for a future PR. Created a ticket here https://dfinity.atlassian.net/browse/XC-243
